netsim/modules/vlan.py

```diff
@@ -391,6 +389,77 @@ def set_node_vlan_mode(n_vlan: Box, node: Box, g_vlan: Box, topology: Box) -> No
                 g_vlan.get('mode',None) or \
                 topology.get('vlan.mode',None) or 'irb'
 
+"""
+populate_node_vlan: merge node-level and topology-level VLAN data
+
+* We're assuming the VLAN needs to be created, so we're not tiptoeing around.
+* Node VLAN mode is evaluated using the precedence principles, everything else
+  is merged
+* _global_merge flag is set once we're done so we're not doing the same thing
+  too many times
+"""
+def populate_node_vlan(vlan: str, node: Box, topology: Box) -> Box:
+  if '_global_merge' in node.vlans[vlan]:
+    return node.vlans[vlan]
+
+  n_vlan = node.vlans[vlan]
+  g_vlan = topology.get('vlans',{}).get(vlan,{})
+
+  n_vlan.mode = n_vlan.mode or \
+                node.get('vlan.mode',None) or \
+                g_vlan.get('mode',None) or \
+                topology.get('vlan.mode',None) or 'irb'
+  n_vlan._global_merge = True
+
+  node.vlans[vlan] = g_vlan + n_vlan
+
+  for m in list(node.vlans[vlan].keys()):                         # Remove irrelevant module parameters
+    if not m in node.module and m in topology.module:             # ... it's safe to use direct references, everyone is using VLAN module
+      node.vlans[vlan].pop(m,None)
+
+  return node.vlans[vlan]
+
+"""
+compute_interface_vlan_mode: Given an interface, a node, and topology, find interface VLAN mode
+"""
+def compute_interface_vlan_mode(intf: Box, node: Box, topology: Box) -> str:
+  vlan = intf.get('vlan.access',None) or intf.get('vlan.native',None)
+  if not vlan:
+    return 'irb'
+  return (
+    intf.get('_vlan_mode',None) or
```

Comment:
Curious if this happens - should only be called once?

Reply:
Didn't check, but it could be that (similar to above) this function is called several times. The "do we already have a result" **get** is just a bit of optimization, the results shouldn't change anyway.